mdapi/intercept_mdapi.cpp

```diff
@@ -0,0 +1,287 @@
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+*/
+
+#include <fstream>
+#include <iomanip>
+
+#include "common.h"
+#include "intercept.h"
+
+#define CL_PROFILING_COMMAND_PERFCOUNTERS_INTEL 0x407F
+
+///////////////////////////////////////////////////////////////////////////////
+//
+static bool convertPropertiesToOCL1_2(
+    const cl_queue_properties* properties,
+    cl_command_queue_properties& ocl1_2_properties )
+{
+    if( properties )
+    {
+        // Convert properties from array of pairs (OCL2.0) to bitfield (OCL1.2)
+        for( int i = 0; properties[ i ] != 0; i += 2 )
+        {
+            if( properties[ i ] == CL_QUEUE_PROPERTIES )
+            {
+                switch( properties[ i + 1 ] )
+                {
+                case CL_QUEUE_PROFILING_ENABLE:
+                    ocl1_2_properties |= CL_QUEUE_PROFILING_ENABLE;
+                    break;
+                case CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE:
+                    ocl1_2_properties |= CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE;
+                    break;
+                default:
+                    return false;
+                }
+            }
+        }
+    }
+
+    return true;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+cl_uint CLIntercept::initCustomPerfCounters()
+{
+    const std::string& metricSetSymbolName = config().DevicePerfCounterCustom;
+    const std::string& metricsFileName = config().DevicePerfCounterFile;
+
+    cl_uint configuration = 0;
+
+    CLI_ASSERT( !metricSetSymbolName.empty() );
+
+    if( m_pMDHelper == NULL )
+    {
+        m_pMDHelper = new MetricsDiscovery::MDHelper();
+        if( m_pMDHelper )
+        {
+            bool initialized = m_pMDHelper->InitMetricsDiscovery(
+                metricSetSymbolName,
+                metricsFileName );
+            if( initialized  )
+            {
+                log( "Metric Discovery initialized.\n" );
+            }
+            else
+            {
+                log( "Metric Discovery failed to initialize.\n" );
+            }
+        }
+    }
+
+    if( m_pMDHelper )
+    {
+        configuration = m_pMDHelper->GetMetricsConfiguration();
+        // TODO: This doesn't appear to be necessary?
+        //m_pMDHelper->SetMetricSetFiltering( API_TYPE_OCL );
+        m_pMDHelper->ActivateMetricSet( );
```

Comment:
ActivateMetricSet should return its result and it should be checked here. Because in case of any error here MDAPI isn't able to continue measurements.